README.md
# GRV - The Generalised RV (GRV) Test

This is an R package version of the standalone GRV R scripts provided at https://wwwf.imperial.ac.uk/~gmontana/grv.html.  The R package was created with minimal efforts such that the package passes `R CMD check --as-cran` using the GRV authors' example script as a package test.

_Disclaimer: I have not intention maintaining this R package beyond my own needs._



NEWS
GRV
===

Version: 0.1.0 [2017-05-04]

 o Minimal R package version of the standalone scripts available from
   https://wwwf.imperial.ac.uk/~gmontana/grv.html as of 2017-02-05.


R/GRV_test.R
# Source: https://wwwf.imperial.ac.uk/~gmontana/software/grv/GRV_test.R

# GRV Test
# ---------

# Date: 21/8/2012
# Author: Christopher Minas

"trace.matrix"<- function(x)
{
	return(sum(diag(x)))
}
"norm.matrix" <- function(mat)
{
	return(sqrt(trace.matrix(mat%*%t(mat))))
}

#' @param n ...
#' @param D ...
#' 
#' @rdname GRV.test
#' @export
"create.G" <- function(n,D) 
{
	D <- as.matrix(D)
	A <- -(1/2)*D^2
      ones <- matrix(data=1,ncol=n,nrow=n)
	centmat <- (diag(n)-ones/n)
      G <- centmat%*%A%*%centmat
      return(G)
}


"perm.info.tr.AB" <- function(A,B,n)
{
	"perm.1st.moment.tr.AB" <- function(A1,B1,n)
	{
		return(A1*B1/(n-1))
	}
	"perm.2nd.moment.tr.AB" <- function(A1,A2,A3,B1,B2,B3,n)
	{
		p11 <- ((n-1)*A3*B3+(B1^2-B3)*(A1^2-A3)+2*(A2-A3)*(B2-B3)+4*A3*B3)/(n*(n-1))
		p22 <- (4*(n-3)*(2*A3-A2)*(2*B3-B2)+2*(2*A3-A1^2)*(2*B3-B1^2)*(n-3)+(2*A2+A1^2-6*A3)*(2*B2+B1^2-6*B3))/(n*(n-1)*(n-2)*(n-3))
		return(p11+p22)
	}
	"perm.3rd.moment.tr.AB" <- function(A1,A2,A3,A4,A5,A6,A7,A8,B1,B2,B3,B4,B5,B6,B7,B8,n)
	{
		n2 <- n^2
		n3 <- n^3
		n4 <- n^4
		return((n2*(n+1)*(n2+15*n-4)*A5*B5 + 4*(n4-8*n3+19*n2-4*n-16)*A6*B6 + 24*(n2-n-4)*(A6*B8+B6*A8)+ 6*(n4-8*n3+21*n2-6*n-24)*A8*B8 + 12*(n4-n3-8*n2+36*n-48)*A7*B7 + 12*(n3-2*n2+9*n-12)*(A1*A3*B7 + A7*B1*B3) + 3*(n4 - 4*n3 - 2*n2+9*n-12)*A1*B1*A3*B3 + 24*((n3 - 3*n2 - 2*n+8)*(A7*B6 + A6*B7) + (n3 - 2*n2 - 3*n+12)*(A7*B8 + A8*B7)) + 12*(n2 - n + 4)*(A1*A3*B6 + B1*B3*A6) + 6*(2*n3 - 7*n2 - 3*n + 12)*(A1*A3*B8 + A8*B1*B3) - 2*n*(n-1)*(n2-n+4)*((2*A6+3*A8)*B5+(2*B6+3*B8)*A5) - 3*n*(n-1)*(n-1)*(n+4)*((A1*A3+4*A7)*B5+(B1*B3+4*B7)*A5) + 2*n*(n-1)*(n-2)*((A1^3 + 6*A1*A2 + 8*A4)*B5+(B1^3 + 6*B1*B2 + 8*B4)*A5) + (A1^3)*((n3-9*n2+23*n-14)*(B1^3)+6*(n-4)*B1*B2+8*B4) + 6*A1*A2*((n-4)*(B1^3)+(n3-9*n2+24*n-14)*B1*B2 + 4*(n-3)*B4) + 8*A4*((B1^3)+3*(n-3)*B1*B2+(n3-9*n2+26*n-22)*B4) - 16*((A1^3)*B6+A6*(B1^3)) - 6*(A1*A2*B6 + A6*B1*B2)*(2*n2-10*n+16) - 8*(A4*B6+A6*B4)*(3*n2-15*n+16)-((A1^3)*B8+A8*(B1^3))*(6*n2-30*n+24)-6*(A1*A2*B8+A8*B1*B2)*(4*n2-20*n+24) - 8*(A4*B8+A8*B4)*(3*n2-15*n+24) - (n-2)*(24*((A1^3)*B7+A7*(B1^3))+6*(A1*A2*B7+A7*B1*B2)*(2*n2-10*n+24)+8*(A4*B7+A7*B4)*(3*n2-15*n+24)+(3*n2-15*n+6)*((A1^3)*B1*B3+A1*A3*(B1^3))+6*(A1*A2*B1*B3+A1*A3*B1*B2)*(n2-5*n+6) + 48*(A4*B1*B3+A1*A3*B4)))/(n*(n-1)*(n-2)*(n-3)*(n-4)*(n-5)))
	}
	Am2 <- A%*%A
	Am3 <- A%*%Am2
	dA <- diag(A)
	Ap2 <- A^2
	Ap3 <- Ap2*A
	A1 <- sum(dA)
	A2 <- trace.matrix(Am2)
	A3 <- trace.matrix(Ap2)
	A4 <- trace.matrix(Am3)
	A5 <- trace.matrix(Ap3)
	A6 <- sum(Ap3)
	A7 <- dA%*%diag(Am2)
	A8 <- dA%*%A%*%dA
	Bm2 <- B%*%B
	Bm3 <- B%*%Bm2
	dB <- diag(B)
	Bp2 <- B^2
	Bp3 <- Bp2*B
	B1 <- sum(dB)
	B2 <- trace.matrix(Bm2)
	B3 <- trace.matrix(Bp2)
	B4 <- trace.matrix(Bm3)
	B5 <- trace.matrix(Bp3)
	B6 <- sum(B^3)
	B7 <- dB%*%diag(Bm2)
	B8 <- dB%*%B%*%dB
	mom1 <- perm.1st.moment.tr.AB(A1, B1, n)
    	mom2 <- perm.2nd.moment.tr.AB(A1,A2,A3,B1,B2,B3,n)
   	mom3 <- perm.3rd.moment.tr.AB(A1,A2,A3,A4,A5,A6,A7,A8,B1,B2,B3,B4,B5,B6,B7,B8,n)
   	mean.T <- mom1
      variance.T <- (mom2 - mom1^2)
      skewness.T <- (mom3 - 3 * variance.T * mom1 - mom1^3)/(variance.T^(3/2))
	return(list(mean.T=mean.T,variance.T=variance.T,skewness.T=skewness.T))
}

#' @param ys ...
#' 
#' @rdname GRV.test
#' @export
"pdf.GRV" <- function(DX,DY,ys)
{
	"pearson.type.3.dgamma" <- function(g,y)
	{
		if(g >= 0){
			ans <- dgamma(y+2/g, shape=4/(g^2), scale = g/2, log = FALSE)
		}else{
			ans <- dgamma(abs(2/g) -y, shape=4/(g^2), scale = abs(g)/2, log = FALSE)
		}
		return(ans)
	}
	DX <- as.matrix(DX)
	DY <- as.matrix(DY)
	n <- nrow(DX)
	GX <- create.G(n,DX)
	GY <- create.G(n,DY)
	constant <- norm.matrix(GX)*norm.matrix(GY)
	perm.info <- perm.info.tr.AB(GX,GY,n)
	constant <- norm.matrix(GX)*norm.matrix(GY)
	m <- perm.info$mean.T
	s <- sqrt(perm.info$variance.T)
	g <- perm.info$skewness.T
	return((constant/s)*sapply(1:length(ys), function(i) pearson.type.3.dgamma(g,(ys[i]*constant - m)/s)))
}

#' @param GX,GY  Two GRV matrices of identical dimensions.
#' @param y ...
#' 
#' @rdname GRV.test
#' @importFrom stats dgamma pgamma
"cdf.GRV" <- function(GX,GY,n,y)
{
	constant <- norm.matrix(GX)*norm.matrix(GY)
	perm.info <- perm.info.tr.AB(GX,GY,n)
	m <- perm.info$mean.T
	s <- sqrt(perm.info$variance.T)
	g <- perm.info$skewness.T
	if(g >= 0){
		ans <- pgamma((y*constant - m)/s+2/g, shape=4/(g^2), scale = g/2)
	}else{
		ans <- pgamma(abs(2/g) -(y*constant - m)/s, shape=4/(g^2), scale = abs(g)/2,lower.tail=FALSE)
	}
	return(ans)
}


#' @rdname GRV.test
#' @export
"GRV" <- function(GX,GY)
{
	return(trace.matrix(GX%*%GY)/(norm.matrix(GX)*norm.matrix(GY)))
}

#' Generalised RV (GRV) test
#'
#' The GRV statistic allows to compare two paired distance matrices.
#'
#' @param DX,DY  Two numeric matrices of identical dimensions to be compared.
#'
#' @example incl/example.R
#'
#' @references
#' Minas C., Curry E., and Montana G., A distance-based test of association between paired heterogeneous genomic data, Bioinformatics, 2013.
#'
#' @rdname GRV.test
#' @export
"GRV.test" <- function(DX,DY) 
{
	DX <- as.matrix(DX)
	DY <- as.matrix(DY)
	n <- nrow(DX)
	GX <- create.G(n,DX)
	GY <- create.G(n,DY)
	grv0 <- GRV(GX,GY)
	p.grv0 <- 1 - cdf.GRV(GX,GY,n,grv0)
	return(c(grv.statistic=grv0,grv.p.value=p.grv0))
}


tests/example.R
## Source: https://wwwf.imperial.ac.uk/~gmontana/software/grv/example.R
library(GRV)

# small GRV example:
# ------------------

# codes required for example:

"rmnorm" <- function (n = 1, mean = rep(0, d), varcov) # returns random multivariate Normal observation
{
    d <- if (is.matrix(varcov))
        ncol(varcov)
    else 1
    z <- matrix(rnorm(n * d), n, d) %*% chol(varcov)
    y <- t(mean + t(z))
    return(y)
}

"rwishart" <- function(df, p) # returns random Wishart matrix
{
    Z <- matrix(0, p, p)
    diag(Z) <- sqrt(rchisq(p, df:(df-p+1)))
    if(p > 1)
    {
        pseq <- 1:(p-1)
        Z[rep(p*pseq, pseq) + unlist(lapply(pseq, seq))] <- rnorm(p*(p-1)/2)
    }
    return(crossprod(Z))
}

"gen.snp.data" <- function(N,Q,rmafs) # returns N samples of Q snps with different minor allele frequencies (mafs)
{
	return(mapply(function(i) sample(c(0,1,2),N,prob=c((1-rmafs[i])^2,2*rmafs[i]*(1-rmafs[i]),rmafs[i]^2),replace=TRUE), 1:Q))
}


"vector.dists" <- function(X,dist) # a selection of distance measures for vectorial observations
{
	"norm.mutual.info.dist" <- function(x1,x2) 
	{
		n <- length(x1)
		nbins <- floor(sqrt(n)) # #choice of nbins as sqrt(n) comes from paper - 'evaluation of gene expression clustering via mutual information distance measure' (2007)
		jnt.density <- hist2d(x1,x2,nbins,show="FALSE")$counts/n
		marginal.density.x1 <- apply(jnt.density,1,sum)
		marginal.density.x2 <- apply(jnt.density,2,sum)
		entropy.x1 <- -sum(ifelse(marginal.density.x1 > 0, marginal.density.x1*log(marginal.density.x1), 0) )
		entropy.x2 <- -sum(ifelse(marginal.density.x2 > 0, marginal.density.x2*log(marginal.density.x2), 0) )
		jnt.density1 <- as.vector(jnt.density)
		jnt.entropy <- -sum(ifelse(jnt.density1 > 0, jnt.density1*log(jnt.density1), 0))
		mi <- entropy.x1 + entropy.x2 -  jnt.entropy
		return(1 - mi/max(entropy.x1,entropy.x2))
	}
	if(dist=="euclidean"){
		n <- nrow(X)
		XXt <- X%*%t(X)
		d <- diag(XXt)
		ones <- matrix(1,nrow=n,ncol=1)
		DX <- sqrt(d%*%t(ones) + ones%*%t(d) - 2*XXt)
	}
	if(dist=="mahalanobis"){
		DX <- sqrt(as.matrix(distance(X,method="mahalanobis")))
	}
	if(dist=="manhattan"){
		DX <- as.matrix(distance(X,method="manhattan"))
	}
	if(dist=="maximum"){
		DX <- as.matrix(dist(X,method="maximum"))
	}
	if(dist=="bray.curtis"){
		DX <- as.matrix(distance(X,method="bray-curtis"))
	}
	if(dist=="correlation.pearson"){
		DX <- 1 - cor(t(X),use="pairwise.complete.obs",method="pearson")
	}
	if(dist=="correlation.spearman"){
		DX <- 1 - cor(t(X),use="pairwise.complete.obs",method="spearman")
	}
	if(dist=="cosine.angle"){
		n <- nrow(X)
		DX <- matrix(0,nrow=n,ncol=n) 
		for(i in 1:(n-1)){
			for(j in 2:n){
				DX[i,j] <- DX[j,i] <-  1- X[i,]%*%X[j,]/(sqrt(sum(X[i,]^2)*sum(X[j,]^2)))
			}
		}
	}
	if(dist=="norm.mutual.information"){ 
		n <- nrow(X)
		DX <- matrix(0,nrow=n,ncol=n)
		for(i in 1:(n-1)){
			for(j in 2:n){
				DX[i,j] <- DX[j,i] <- norm.mutual.info.dist(X[i,],X[j,])	
			}
		}
	}
	return(DX)
}

"snp.dists" <- function(data,method) # snp distances taken from 'Similarity Measures for Clustering SNP Data' (2005)
{
	n <- nrow(data)
	vec.pairs <- combn(c(1:n),2)
	no.vec.pairs <- ncol(vec.pairs)
	dmat <- matrix(0,nrow=n,ncol=n)
	if(method=="IBS"){
		sim.vals <- sapply(1:no.vec.pairs, function(i) 0.5*mean(2 - abs(data[vec.pairs[1,i],] - data[vec.pairs[2,i],])))      
		dmat[lower.tri(dmat)] <- 1-sim.vals
		dmat <- dmat+t(dmat)
	}
	if(method=="simple.match"){
		sim.vals <- sapply(1:no.vec.pairs, function(i) mean(data[vec.pairs[1,i],] - data[vec.pairs[2,i],]==0))      
		dmat[lower.tri(dmat)] <- 1-sim.vals
		dmat <- dmat+t(dmat)
	}
	if(method=="sokal.sneath"){
		p <- ncol(data)
		sim.vals <- sapply(1:no.vec.pairs, function(i) {m.plus <- sum(data[vec.pairs[1,i],] - data[vec.pairs[2,i],]==0)
			 							m.minus <- p - m.plus
										m.plus/(m.plus+0.5*m.minus)})
		dmat[lower.tri(dmat)] <- 1-sim.vals
		dmat <- dmat+t(dmat)
	}
	if(method=="rogers.tanimoto.1"){
		p <- ncol(data)
		sim.vals <- sapply(1:no.vec.pairs, function(i) {m.plus <- sum(data[vec.pairs[1,i],] - data[vec.pairs[2,i],]==0)
										m.minus <- p - m.plus
										m.plus/(m.plus+2*m.minus)})
		dmat[lower.tri(dmat)] <- 1-sim.vals
		dmat <- dmat+t(dmat)
	}
	if(method=="hamman.1"){
		p <- ncol(data)
		sim.vals <- sapply(1:no.vec.pairs, function(i) {m.plus <- sum(data[vec.pairs[1,i],] - data[vec.pairs[2,i],]==0)
										m.minus <- p - m.plus
										(m.plus-m.minus)/p})
		sim.vals2 <- sim.vals + abs(min(sim.vals))
		dmat[lower.tri(dmat)] <- 1-(sim.vals2/max(sim.vals2))
		dmat <- dmat+t(dmat)
	}
	return(dmat)
}

library(MASS) 
library(graphics)
library(stats)
library(Matrix)
library(combinat)
library(ecodist)
library(gplots)

set.seed(101)

# First we demonstrate how to use the GRV test:
# generate (1) X = N x P multivariate data matrix
#          (2) Y = N x Q snp data matrix
#	     (3) obtain corresponding N x N distance matrices DX and DY and apply GRV test

# simulate X:

N <- 50
P <- 10
X <- rmnorm(N, mean=runif(P,-6,6), varcov=rwishart(df=P,p=P))	

# simulate Y:

Q <- 5
Y <- gen.snp.data(N,Q,rmafs=runif(Q,0.1,0.5))

# apply mahalanobis distance to X, and IBS distance to Y, and apply GRV test:

DX.mah <- vector.dists(X,dist="mahalanobis")
DY.ibs <- snp.dists(Y,method="IBS")

GRV.test(DX.mah,DY.ibs)
#grv.statistic   grv.p.value 
#   0.14236262    0.03878647 

# plot approximate probability density function (PDF) of GRV against sampling distribution using 10^4 Monte Carlo permutations:

GX <- create.G(N,DX.mah)
GY <- create.G(N,DY.ibs)

ys <- seq(0,0.2,len=1000)
pdf.GRV.vals <- pdf.GRV(DX.mah,DY.ibs,ys)
GRV.perms <- sapply(1:10000, function(i) {perms <- sample(1:N,replace=FALSE)
							GRV(GX[perms,perms],GY)})
plot(ys,pdf.GRV.vals,type="l",ylab="PDF(y)",xlab="y",main="PDF of GRV")
hist(GRV.perms,prob=TRUE,add=TRUE)



incl/example.R
# small GRV example:
# ------------------

# codes required for example:

"rmnorm" <- function (n = 1, mean = rep(0, d), varcov) # returns random multivariate Normal observation
{
    d <- if (is.matrix(varcov))
        ncol(varcov)
    else 1
    z <- matrix(rnorm(n * d), n, d) %*% chol(varcov)
    y <- t(mean + t(z))
    return(y)
}

"rwishart" <- function(df, p) # returns random Wishart matrix
{
    Z <- matrix(0, p, p)
    diag(Z) <- sqrt(rchisq(p, df:(df-p+1)))
    if(p > 1)
    {
        pseq <- 1:(p-1)
        Z[rep(p*pseq, pseq) + unlist(lapply(pseq, seq))] <- rnorm(p*(p-1)/2)
    }
    return(crossprod(Z))
}

"gen.snp.data" <- function(N,Q,rmafs) # returns N samples of Q snps with different minor allele frequencies (mafs)
{
	return(mapply(function(i) sample(c(0,1,2),N,prob=c((1-rmafs[i])^2,2*rmafs[i]*(1-rmafs[i]),rmafs[i]^2),replace=TRUE), 1:Q))
}


"vector.dists" <- function(X,dist) # a selection of distance measures for vectorial observations
{
	"norm.mutual.info.dist" <- function(x1,x2) 
	{
		n <- length(x1)
		nbins <- floor(sqrt(n)) # #choice of nbins as sqrt(n) comes from paper - 'evaluation of gene expression clustering via mutual information distance measure' (2007)
		jnt.density <- hist2d(x1,x2,nbins,show="FALSE")$counts/n
		marginal.density.x1 <- apply(jnt.density,1,sum)
		marginal.density.x2 <- apply(jnt.density,2,sum)
		entropy.x1 <- -sum(ifelse(marginal.density.x1 > 0, marginal.density.x1*log(marginal.density.x1), 0) )
		entropy.x2 <- -sum(ifelse(marginal.density.x2 > 0, marginal.density.x2*log(marginal.density.x2), 0) )
		jnt.density1 <- as.vector(jnt.density)
		jnt.entropy <- -sum(ifelse(jnt.density1 > 0, jnt.density1*log(jnt.density1), 0))
		mi <- entropy.x1 + entropy.x2 -  jnt.entropy
		return(1 - mi/max(entropy.x1,entropy.x2))
	}
	if(dist=="euclidean"){
		n <- nrow(X)
		XXt <- X%*%t(X)
		d <- diag(XXt)
		ones <- matrix(1,nrow=n,ncol=1)
		DX <- sqrt(d%*%t(ones) + ones%*%t(d) - 2*XXt)
	}
	if(dist=="mahalanobis"){
		DX <- sqrt(as.matrix(distance(X,method="mahalanobis")))
	}
	if(dist=="manhattan"){
		DX <- as.matrix(distance(X,method="manhattan"))
	}
	if(dist=="maximum"){
		DX <- as.matrix(dist(X,method="maximum"))
	}
	if(dist=="bray.curtis"){
		DX <- as.matrix(distance(X,method="bray-curtis"))
	}
	if(dist=="correlation.pearson"){
		DX <- 1 - cor(t(X),use="pairwise.complete.obs",method="pearson")
	}
	if(dist=="correlation.spearman"){
		DX <- 1 - cor(t(X),use="pairwise.complete.obs",method="spearman")
	}
	if(dist=="cosine.angle"){
		n <- nrow(X)
		DX <- matrix(0,nrow=n,ncol=n) 
		for(i in 1:(n-1)){
			for(j in 2:n){
				DX[i,j] <- DX[j,i] <-  1- X[i,]%*%X[j,]/(sqrt(sum(X[i,]^2)*sum(X[j,]^2)))
			}
		}
	}
	if(dist=="norm.mutual.information"){ 
		n <- nrow(X)
		DX <- matrix(0,nrow=n,ncol=n)
		for(i in 1:(n-1)){
			for(j in 2:n){
				DX[i,j] <- DX[j,i] <- norm.mutual.info.dist(X[i,],X[j,])	
			}
		}
	}
	return(DX)
}

"snp.dists" <- function(data,method) # snp distances taken from 'Similarity Measures for Clustering SNP Data' (2005)
{
	n <- nrow(data)
	vec.pairs <- combn(c(1:n),2)
	no.vec.pairs <- ncol(vec.pairs)
	dmat <- matrix(0,nrow=n,ncol=n)
	if(method=="IBS"){
		sim.vals <- sapply(1:no.vec.pairs, function(i) 0.5*mean(2 - abs(data[vec.pairs[1,i],] - data[vec.pairs[2,i],])))      
		dmat[lower.tri(dmat)] <- 1-sim.vals
		dmat <- dmat+t(dmat)
	}
	if(method=="simple.match"){
		sim.vals <- sapply(1:no.vec.pairs, function(i) mean(data[vec.pairs[1,i],] - data[vec.pairs[2,i],]==0))      
		dmat[lower.tri(dmat)] <- 1-sim.vals
		dmat <- dmat+t(dmat)
	}
	if(method=="sokal.sneath"){
		p <- ncol(data)
		sim.vals <- sapply(1:no.vec.pairs, function(i) {m.plus <- sum(data[vec.pairs[1,i],] - data[vec.pairs[2,i],]==0)
			 							m.minus <- p - m.plus
										m.plus/(m.plus+0.5*m.minus)})
		dmat[lower.tri(dmat)] <- 1-sim.vals
		dmat <- dmat+t(dmat)
	}
	if(method=="rogers.tanimoto.1"){
		p <- ncol(data)
		sim.vals <- sapply(1:no.vec.pairs, function(i) {m.plus <- sum(data[vec.pairs[1,i],] - data[vec.pairs[2,i],]==0)
										m.minus <- p - m.plus
										m.plus/(m.plus+2*m.minus)})
		dmat[lower.tri(dmat)] <- 1-sim.vals
		dmat <- dmat+t(dmat)
	}
	if(method=="hamman.1"){
		p <- ncol(data)
		sim.vals <- sapply(1:no.vec.pairs, function(i) {m.plus <- sum(data[vec.pairs[1,i],] - data[vec.pairs[2,i],]==0)
										m.minus <- p - m.plus
										(m.plus-m.minus)/p})
		sim.vals2 <- sim.vals + abs(min(sim.vals))
		dmat[lower.tri(dmat)] <- 1-(sim.vals2/max(sim.vals2))
		dmat <- dmat+t(dmat)
	}
	return(dmat)
}

library(MASS) 
library(graphics)
library(stats)
library(Matrix)
library(combinat)
library(ecodist)
library(gplots)

set.seed(101)

# First we demonstrate how to use the GRV test:
# generate (1) X = N x P multivariate data matrix
#          (2) Y = N x Q snp data matrix
#	     (3) obtain corresponding N x N distance matrices DX and DY and apply GRV test

# simulate X:

N <- 50
P <- 10
X <- rmnorm(N, mean=runif(P,-6,6), varcov=rwishart(df=P,p=P))	

# simulate Y:

Q <- 5
Y <- gen.snp.data(N,Q,rmafs=runif(Q,0.1,0.5))

# apply mahalanobis distance to X, and IBS distance to Y, and apply GRV test:

DX.mah <- vector.dists(X,dist="mahalanobis")
DY.ibs <- snp.dists(Y,method="IBS")

GRV.test(DX.mah,DY.ibs)
#grv.statistic   grv.p.value 
#   0.14236262    0.03878647 

# plot approximate probability density function (PDF) of GRV against sampling distribution using 10^4 Monte Carlo permutations:

GX <- create.G(N,DX.mah)
GY <- create.G(N,DY.ibs)

ys <- seq(0,0.2,len=1000)
pdf.GRV.vals <- pdf.GRV(DX.mah,DY.ibs,ys)
GRV.perms <- sapply(1:10000, function(i) {perms <- sample(1:N,replace=FALSE)
							GRV(GX[perms,perms],GY)})
plot(ys,pdf.GRV.vals,type="l",ylab="PDF(y)",xlab="y",main="PDF of GRV")
hist(GRV.perms,prob=TRUE,add=TRUE)

